Add auctions to NIP-15.

ibz opened the pull request.

ibz commented on 15.md:
```diff
@@ -243,6 +247,110 @@ Create a customized user experience using the `naddr` from [NIP-19](https://gith
+
+Another thing that can happen is - if bids happen very close to the end date of the auction - for the merchant to decide to extend the auction duration for a few more minutes. This is done by passing a `duration_extended` field as part of a bid confirmation, which would contain a number of seconds by which the initial duration is extended. So the actual end date of an auction is always `start_dat ...
+
+Auctions can also have a "reserve bid" - which is an amount unknown to anyone else (therefore not part of the auction event and remembered by the merchant admin software) - and represents the minimum amount for which the seller is willing to sell the item. Anything less than that and the item will not be sold, even if the auction ends and there is a top bidder! The marketplace client software has ...
```
"reserve bid" is pretty standard in auctions. Sellers actually want to have that because ... if they would set very high "start bid", they might not get any bids at all (buyers get scared away).

But I am happy to remove these and reduce NIP complexity. In fact the implementation can work either way - having these in the NIP or not. It's still up to the merchant how the winner is picked. So yes, let's keep it simple!

ibz commented on 15.md:
```diff
@@ -243,6 +247,110 @@ Create a customized user experience using the `naddr` from [NIP-19](https://gith
+
+NB: items sold as an auction are very similar in structure to fixed-price items, with some important differences worth noting.
+
+* While fixed-price items can be priced in any currency and the conversion to Bitcoin happens on checkout, bids on items sold as an auction happen in sats. Client implementations will probably display conversions to various fiat currencies, but the bidder always commits to an amount to be paid in sats if they will happen to be the top bidder at the end of the auction.
```
I think other currencies *might* work, but it would be pretty strange, because the payment is always in sats.

For fixed-price items it is good, because "what you see is what you get" - you see 50$ and the amount of sats, you click "buy" and you pay that amount of sats.

For auctions it is strange. Say I commit to buying this item for 50$ (= x sats) (remember, a bid is a commitment!) - but one week later, when I need to actually pay, I realize that I don't have enough sats in my wallet. So I have no way of paying what I committed to pay!

I am happy to not have this limitation in the NIP though - and if other implementations choose to allow other currencies, they would be free to do so.

How would you do that? Specify a `bid_currency` on the auction object and then all amounts would be in that currency? Sounds simple enough and worth it for keeping the NIP as general as possible!

motorina0 replied:
> How would you do that? Specify a bid_currency on the auction object

The `auction object` is a particular type of product. This means it belongs to a `Stall` and the `Stall` has a `currency` that can be used.

ibz replied:
Interesting, but does that mean I need a separate stall if I want to run auctions? I can't have auctions and fixed-price items on the same stall?

motorina0 replied:
> I can't have auctions and fixed-price items on the same stall?
- yes, it should possible

ibz commented on 15.md:
```diff
@@ -243,6 +247,110 @@ Create a customized user experience using the `naddr` from [NIP-19](https://gith
+
+The `content` field is a JSON which includes *at least* a `status`, which is one of `accepted`, `rejected`, `pending`, or `winner`. `winner` is how the *winning bid* is replied to after the auction ends and the winning bid is picked by the merchant.
+
+The reasons for which a bid can be marked as `rejected` or `pending` are up to the merchant's implementation and configuration - they could be anything from basic validation errors (amount too low) to the bidder being blacklisted or to the bidder lacking sufficient *trust*, which could lead to the bid being marked as `pending` until sufficient verification is performed. The difference between the ...
```
"Merchants verify bidders" by (possibly) asking for a deposit is - sort of - what we do, amongst others! We call it "skin in the game", which is a Nostr (NIP-58) badge you can purchase. You can totally win an auction without revealing your real identity.

All in all, this is probably the most complex issue when it comes to Nostr auctions and I think we are too early to set it in stone at the protocol level. We might to it later, as we learn more and gather more feedback.

But mentioning a "deposit" as part of the NIP is - I feel - encouraging implementations to be money handlers, which is a dangerous path to go. Unless you say that the deposits stay in the merchant's hands. But then - we need another protocol to handle the refunds.

ibz replied:
We are believers in "reputation by participation". Real world auctions don't (really) allow this, but Nostr does. You can be a keypair with nothing else known about you, just a history of having successfully bought stuff in the past. And the history is known by the marketplace.

The issue really is newcomers. Fresh keypairs need *some* way to acquire trust. Known keypairs are fine, because the more reputation you have, the more likely you want to hold on to that and not lose it by screwing up and have your whole reputation slashed!

That's why the decision is not really done at the protocol level, but rather at the implementation level. How can you define what is enough "skin in the game" that makes you unlikely to want to be a troll?

ibz commented:
BTW, our implementation of NIP-15 (with fixed price items *and* auctions) is now live at plebeian.market!

ibz commented:
I addressed most comments from @motorina0. @fiatjaf what are the next steps for getting this merged? Anyone else needs to review?

BTW, our implementation of NIP-15 (with fixed price items and auctions) is now live at plebeian.market!
